src/lib/zork/zork.ts

```diff
@@ -0,0 +1,378 @@
+    console.info(`[get] icecandidate from client ${ctx.clientId}`);//: ${msg}`);
+    ctx.peerConn.addIceCandidate(parsed.candidate);
+  } else if (parsed.type === 'offer') {
+    console.info(`[get] got offer`);//: ${msg}`);
+    ctx.peerConn.setRemoteDescription(parsed);
+    ctx.peerConn.createAnswer((answer: any) => {
+      const answerJson = JSON.stringify(answer);
+      console.info(`[get] created answer`);//: ${answerJson}`);
+      ctx.peerConn.setLocalDescription(answer);
+      ctx.socket.write(`${JSON.stringify(answer)}\n`);
+    }, console.error);
+  } else {
+    console.error(`[get] unexpected msg: ${msg}`);
+  }
+};
+
+const msgHandlerByMode: {[mode: string]: (ctx: Context, msg: string) => void} = {
+  'give': handleMsgForModeGive,
+  'get': handleMsgForModeGet
+};
+
+const handleMsg = (ctx: Context, msg: string) => {
+  if (ctx.mode) {
+    // Already in 'give' or 'get' mode. Dispatch to corresponding message handler.
+    const msgHandler = msgHandlerByMode[ctx.mode];
+    if (msgHandler) {
+      msgHandler(ctx, msg);
+    } else {
+      console.error(`no message handler for mode: ${ctx.mode}`);
+    }
+  } else {
+    // Not yet in 'give' or 'get' mode. Treat msg as command.
+    const cmd = parseCmd(msg);
+    const cmdHandler = cmdHandlerByCmdId[cmd.id] || handleCmdInvalid;
+    cmdHandler(ctx, cmd);
+  }
+};
+
+const zorkServer = net.createServer((client) => {
+
+  const ctx: Context = {
+    clientId: nconnectionsMade++,
+    socket: client,
+    mode: null,
+    transformer: null,
+    peerConn: null
+  };
+
+  console.info(`client ${ctx.clientId} connected`);
+
+  // Handle receiving data from this client.
+  // Buffer for partially transmitted messages.
+  let buffer = '';
+
+  client.on('data', (data) => {
+    const chunk = data.toString();
+    const msgs = chunk.split(MSG_DELIM_RE);
+    const msgDelimNotFound = msgs.length === 1;
+    if (msgDelimNotFound) {
+      // No delimiter found means we only have part of a message. Continue
+      // adding it to `buffer` (potentially in subsequent callbacks as well)
+      // until we do reach a delimiter.
+      buffer += chunk;
+      return;
+    }
+    // Message delimiter found. After adding any message parts we buffered
+    // previously, we should now have at least one complete message.
+    msgs[0] = buffer + msgs[0];
+    // If the data we've read off the socket ended with the message delimiter,
+    // the last element of `msgs` is empty string.
+    // Otherwise, the last element is only part of a message which we will have
+    // to reconstruct in a future callback once more data is available.
+    // Popping off this last element and setting `buffer` to it handles both
+    // cases.
+    buffer = msgs.pop();
+    // Process the complete messages we've received.
+    for (let msg of msgs) {
+      handleMsg(ctx, msg);
+    }
+  });
+
+  client.on('end', () => {
+    console.info(`client ${ctx.clientId} disconnected`);
+    if (ctx.mode === 'give') {
```

Comment:
So, I meant active peer connectios. Currently zork closes the connection to port 9000 as soon as the peerconnection is started (or fails).